### DISCUSSION TITLE
Augmenting UA nodes with custom Python properties

### DISCUSSION THREAD

**Discussion by xmas79** [FreeOpcUa/opcua-asyncio#665] [Q&A] [answered]
Hello,
I have a small question... I want to augment the UA nodes inside a server with custom properties (python side). The idea is to write the server model in the GUI and exploit the references created from a GUI. I'm going to have a very complex model, importing these references from XML is a great help.

I have got a minimal working example below, but the problem with it is that manual bookmarking of all the nodes is needed. In the example, I create a simple server with a single object node. This node is "augmented" with a custom python property that can be used if I use the original object, but I don't know how to get the "augmented" object from the server. Is that possible?

Thanks.

```
import asyncio
from asyncua import Server

async def main():
    server = Server()

    await server.init()
    server.set_endpoint('opc.tcp://0.0.0.0:4840/freeopcua/server/')

    uri = 'http://examples.freeopcua.github.io'
    idx = await server.register_namespace(uri)

    myobj = await server.nodes.objects.add_object(idx, 'MyObject')
    setattr(myobj, "hack", "My Value")
    print(myobj.hack) # That's fine
    
    myobj2 = await server.nodes.objects.get_child([ f"{idx}:MyObject"])
    print(myobj2.hack) # This fails with AttributeError: 'Node' object has no attribute 'hack' --> myobj2 is a plain Node object

    async with server:
        while True:
            await asyncio.sleep(1)

if __name__ == "__main__":
    asyncio.run(main())

```

**Comment by AndreasHeine** [comment]
```python
print(myobj2.hack) # This fails with AttributeError: 'Node' object has no attribute 'hack' --> myobj2 is a plain Node object
```

for sure it creates a new nodeclass instance which dont has that attr

**Reply by xmas79** [reply]
And that was unexpected to me... Is there a chance to get the "original" node? I would expect the server holds a copy of the original node somewhere... Manual bookmarking seems redundant to me

**Reply by AndreasHeine** [reply]
the node-class is just a wrapper class it does not get stored with methods! check out the adressspace.py for insight.

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/master/asyncua/server/address_space.py

**Comment by xmas79** [comment, marked as answer]
I understand, the idea is to return only node identifiers (ns=x;i=y) and never expose the internal nodes (I see the internal database is updated here `https://github.com/FreeOpcUa/opcua-asyncio/blob/2f0c9fc1d40252e08e64242ba0f65465dd077feb/asyncua/server/address_space.py#L280`). I will manual bookmark the nodes as a simple python dictionary (eg. `mynodes[myobj.nodeid.to_string()] = myobj`).

Thanks!
